worker/lease/dead_manager_test.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package lease_test
+
+import (
+	"reflect"
+
+	gc "gopkg.in/check.v1"
+	"gopkg.in/juju/worker.v1"
+
+	"github.com/juju/juju/worker/lease"
+)
+
+type DeadManagerSuite struct{}
+
+var _ = gc.Suite(&DeadManagerSuite{})
+
+type DeadManagerError struct{}
+
+const DeadManagerErrorMessage = "DeadManagerError"
+
+func (*DeadManagerError) Error() string {
+	return DeadManagerErrorMessage
+}
+
+// This creates a new DeadManager and calls all of its exported methods with zero
+// values. All methods should return the specified DeadManagerError.
+func (s *DeadManagerSuite) TestDeadManager(c *gc.C) {
+	deadManagerErr := DeadManagerError{}
+	deadManager := lease.NewDeadManager(&deadManagerErr)
+	deadManagerType := reflect.TypeOf(deadManager)
+	deadManagerValue := reflect.ValueOf(deadManager)
+	errorIface := reflect.TypeOf((*error)(nil)).Elem()
+	workerIface := reflect.TypeOf((*worker.Worker)(nil)).Elem()
+
+	for i := 0; i < deadManagerType.NumMethod(); i++ {
+		method := deadManagerType.Method(i)
+		methodV := deadManagerValue.MethodByName(method.Name)
+
+		var args []reflect.Value
+		for n := 0; n < methodV.Type().NumIn(); n++ {
+			argType := methodV.Type().In(n)
+			args = append(args, reflect.Zero(argType))
+		}
+
+		for j := 0; j < method.Type.NumOut(); j++ {
+			if returnType := method.Type.Out(j); returnType.Implements(errorIface) {
+				errorValue := methodV.Call(args)[j]
+				if _, ok := workerIface.MethodByName(method.Name); ok {
```

Comment:
Actually, we really just need to test that the worker Wait() method returns the dead manager error message.
And that the methods on the Claimer and Checker interfaces, ie Claim(), WaitUntilExpired(), Token(), return the "lease manager stopped" error.
There's a lot of reflection code here that makes it a little hard for the casual reader to grok. I think we can just simplify this test down to

deadManager := lease.NewDeadManager(&deadManagerErr)
c.Assert(deadManager.Wait(), gc.ErrorMatches, deadManagerErrorMessage)
c.Assert(deadManager.Claim("", "", time.Second), gc.ErrorMatches, leaseManagerStoppedMessage)
c.Assert(deadManager.WaitUntilExpired("", ""), gc.ErrorMatches, leaseManagerStoppedMessage)
c.Assert(deadManager.Check(nil), gc.ErrorMatches, leaseManagerStoppedMessage)